src/Features/LanguageServer/Protocol/Handler/RequestContext.cs

```diff
@@ -14,6 +16,122 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     /// </summary>
     internal readonly struct RequestContext
     {
+        public static RequestContext Create(
+            TextDocumentIdentifier? textDocument,
+            string? clientName,
+            ClientCapabilities clientCapabilities,
+            ILspWorkspaceRegistrationService lspWorkspaceRegistrationService,
+            Dictionary<Workspace, (Solution workspaceSolution, Solution lspSolution)>? solutionCache,
+            IDocumentChangeTracker? documentChangeTracker)
+        {
+            // This method asks the solution provider for a solution, and then updates it based on the state of the world
+            // as we know it. You may look at this and think "why doesn't the solution provider keep track of the state
+            // then it can just provide the right solutions" and at first glance that is appealing, but there are two big
+            // benefits to this queue tracking documents separately:
+            //
+            // 1. Since this queue manages the execution of handlers, we can ensure that the document state tracker
+            //    is only used from mutating request handlers, or outside of any handlers, and therefore we know
+            //    that calls to it cannot overlap, and it doesn't have to worry about threading.
+            //
+            // 2. We specifically only want the updated solution at the start of a request. If the solution provider
+            //    handed out "the right" solution, then a request could ask it for something at the start of processing
+            //    and at the end of processing, and get different results each time.
+
+            // Assume the first workspace registered is the main one
```

Comment:
in a followup we should take a look at using potentially all workspaces for requests w/out a document (e.g workspace symbols)

Reply:
We did talk about this briefly at one point, but there are a lot of questions this brings up that I don't have clear answers to. Do we call all workspaces, and somehow combine their results? Or do we call them all until we get a valid response? And what happens if one request goes to workspace 2, but then something changes and the next request ends up looking at workspace 1, or 3?

Reply:
Right I had forgotten we discussed this.  To be fair it was last year! ;)